src/connector/src/source/base.rs

```diff
@@ -450,12 +450,12 @@ impl SplitMetaData for SplitImpl {
         Self::restore_from_json_inner(&split_type, inner_value.into())
     }
 
-    fn update_with_offset(&mut self, start_offset: String) -> Result<()> {
+    fn update_offset(&mut self, last_seen_offset: String) -> Result<()> {
```

Comment:
Why name it `last_seen_offset`?  I think a general `offset` is suitable, since this interface is just to update the offset of the in-memory split.

Reply:
Because the offset is actually the "last seen offset". The problem is that such `XxxSplit`s are also used to recover a reader. When recovering, the concept of `offset` is changed to "please read messages from here". As you can see the `XxxSplit`s need to correctly handle/represent such concept difference.

Reply:
Can check the follow-up PR https://github.com/risingwavelabs/risingwave/pull/16257 that fixes the concept mismatch in `KafkaSplit`

Reply:
IMO we should `+1` if possible when `update_offset` (unfortunately some source platforms don't allow simply doing addition on offset), that's in my mind the most elegant and clear solution. But in our current actual implementation, we have 3 cases:

1. `+1` when creating reader (Kafka)
2. Set offset to `AfterSequenceNumber(last_seen_offset)` (Kinesis)
3. Read from `last_seen_offset` on recovery, meaning duplicated items (Debezium CDC). The duplication seem unavoidable, see https://debezium.io/documentation/reference/stable/development/engine.html#_handling_failures.

Reply:
Yeah, in cdc scenario we don't have API to do this `+1` (logically) operation for offset.

Reply:
IIUC, for the cases of CDC, the `offset` actually being used is the **next** offset instead of last offset.

Keep in mind that not system use continuous offset. CDC i.e. the WAL of databases is one of the example, because the offset is simply the file offset in WAL file, and the log entries have various lengths. In this case, keeping the offset of last-seen log entry doesn't make sense to me.

Additionally, how to represent a zero offset in the `last_seen_offset` way? `-1`? This is the case when a new partition or topic was just created.

Reply:
<img width="1513" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/14017336/dc5cbf29-f975-4285-af62-ee680855f440">

may take a look at this example, the behavior of recording the very beginning is writing nothing.